Add support for XML clips in chains.

bmatherly opened the pull request:
This allows speed effects to XML clips.
It also adds a properties panel for XML clips.

Also fixes this issue:
https://forum.shotcut.org/t/issues-when-using-mlt-as-a-clip-in-a-timeline/44164

If the XML clip has a filter on the output, the filter will not appear in the project - it is "hidden" inside the chain.

Depends on
https://github.com/mltframework/mlt/pull/1047

This is what the properties panel looks like (similar to the Output properties):
![image](https://github.com/user-attachments/assets/a25bd3aa-260b-4bd5-bdf8-edd66a4b1bd2)

The way this is designed, the embedded XML will always be rendered at the XML clip profile (by the consumer producer). Then it will be converted to the project profile if necessary.

Discussion: Should we show a warning in the properties panel if the XML clip profile does not match the project profile?

Discussion: Do we want to add a button or method to lauch Shotcut from the properties panel so the clip can be edited in another instance of Shotcut? Maybe that would be confusing to have two instances at the same time.

bmatherly commented:
Of interest:
https://forum.shotcut.org/t/adding-to-timeline-a-mlt-as-a-clip-with-different-fps-crashes-shotcut/46418

An idea from that post is that I could add the duration to the properties panel.

ddennedy commented:
> An idea from that post is that I could add the duration to the properties panel.

Some things I noticed while debugging the forum report and testing these branches:
- The audio plays with heavy crackling artifacts if you nest a project with audio and a different frame rate (outer project 60 fps, inner project 29.97).
- dragging .mlt file to playlist, double-click to open from playlist, Properties does not show the new producer widget
- same goes when drag .mlt to Timeline
- File > Open MLT XML as Clip (before adding to Timeline) is giving me empty Properties (I tried a couple projects)

![image](https://github.com/user-attachments/assets/974aeccf-2d76-4f10-a187-9501cdf3360f)

I am using the MLT branch as well.

ddennedy commented:
FYI, the audio issue on mismatching frame rate exists regardless this change, but if this change can accommodate speed I wondered if it could improve this. It is worse when the nested project fps is less than the outer's.

bmatherly commented:
> FYI, the audio issue on mismatching frame rate exists regardless this change, but if this change can accommodate speed I wondered if it could improve this. It is worse when the nested project fps is less than the outer's.

I will look into this. My intention is to fix these kinds of problems.

> It permits the addition of Speed: Forward & Reverse and Time Remap, which can go in reverse, and the project might perform very poorly because it masks whether its videos seek well (has no B frames).

This is a very good point. Part of my motivation was to allow speed effects on a nested clip. But now I think that is not a good idea and probably not that important since a user can simply encode the embedded project and apply speed to the encoded file. I think I will remove the chain wrapper (but keep the producer_consumer wrapper).

Thanks again for your testing. I will spend some more time on this. I would like it to be bullet proof before we show it to a wider audience.

bmatherly commented:
> Discussion: Should we show a warning in the properties panel if the XML clip profile does not match the project profile?

This is done

> The audio plays with heavy crackling artifacts if you nest a project with audio and a different frame rate (outer project 60 fps, inner project 29.97).

Fixed
    
> dragging .mlt file to playlist, double-click to open from playlist, Properties does not show the new producer widget

Fixed

> same goes when drag .mlt to Timeline

Fixed

> File > Open MLT XML as Clip (before adding to Timeline) is giving me empty Properties (I tried a couple projects)

Fixed

> It permits the addition of Speed: Forward & Reverse and Time Remap, which can go in reverse, and the project might perform very poorly because it masks whether its videos seek well (has no B frames).

I did not prohibit speed effects (but I can easily if we want to). There are some use cases that I would like to support. For example, adding a timer on a clip and then adding that as a subproject to show a timer moving in fast motion with a fast forward video. I am open minded about this.

The major difference with this latest version is that the xml-clip producer is always opened in its native profile (frame rate) and a timeremap link is added to the chain if the frame rate does not match the project frame rate. The timeremap link performs frame rate conversion.

The only thing I am slightly disatisfied with is that I notice interlaced files have lower performance due to two deinterlace links being stacked (one inside the xml producer and one in the encapsulating chain). Deinterlace links cause extra processing even when the project is progressive because they pull previous/next frames in case a filter requests deinterlace. I haven't thought of a good solution for this, and I am not sure of the priority.

ddennedy commented:
> I did not prohibit speed effects (but I can easily if we want to). 

What about only allowing Speed: Forward Only? Basically, check `seekReverse` against the producer's `mlt_service`. It would be ideal if we could filter the filter menu based on this, but I doubt we can do that yet.

bmatherly commented:
> It would be ideal if we could filter the filter menu based on this, but I doubt we can do that yet.

I did it.
